src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts

```diff
@@ -0,0 +1,224 @@
+    public actionLabels: ActionLabelsI18n
+  ) {
+    this.permission = this.authStorageService.getPermissions().rgw;
+    const createRealmAction: CdTableAction = {
+      permission: 'create',
+      icon: Icons.add,
+      name: this.actionLabels.CREATE_REALM,
+      click: () => this.openModal('realm')
+    };
+    const createZonegroupAction: CdTableAction = {
+      permission: 'create',
+      icon: Icons.add,
+      name: this.actionLabels.CREATE_ZONEGROUP,
+      click: () => this.openModal('zonegroup')
+    };
+    const createZoneAction: CdTableAction = {
+      permission: 'create',
+      icon: Icons.add,
+      name: this.actionLabels.CREATE_ZONE,
+      click: () => this.openModal('zone')
+    };
+    this.createTableActions = [createRealmAction, createZonegroupAction, createZoneAction];
+  }
+
+  openModal(entity: any, edit = false) {
+    const entity_name = edit ? entity.data.name : entity;
+    const action = edit ? 'edit' : 'create';
+    const initialState = {
+      resource: entity_name,
+      action: action,
+      info: entity
+    };
+    this.bsModalRef = this.modalService.show(RgwMultisiteFormComponent, initialState, {
+      size: 'lg'
+    });
+  }
+
+  ngOnInit() {
+    this.sub = this.timerService
+      .get(() => this.rgwDaemonService.getMultisiteInfo(), 10000)
+      .subscribe((data: any) => {
+        this.loadingIndicator = false;
+        this.nodes = this.abstractTreeData(data);
+      });
+  }
+
+  ngOnDestroy() {
+    this.sub.unsubscribe();
+  }
+
+  getRealmTree(realm: any) {
+    let rootNodes = {};
+    rootNodes['id'] = realm.id;
+    this.realmIds.push(realm.id);
+    rootNodes['name'] = realm.name + ' (realm)';
+    rootNodes['info'] = realm;
+    rootNodes['is_default'] = realm.name === this.defaultsInfo['default_realm'] ? true : false;
+    return rootNodes;
+  }
+
+  getZonegroupTree(zonegroup: RgwZonegroup, realm?: RgwRealm) {
+    let firstChildNodes = {};
+    firstChildNodes['id'] = zonegroup.id;
+    firstChildNodes['name'] = zonegroup.name + ' (zonegroup)';
+    firstChildNodes['info'] = zonegroup;
+    firstChildNodes['is_default'] =
+      zonegroup.name === this.defaultsInfo['default_zonegroup'] ? true : false;
+    firstChildNodes['is_master'] = zonegroup.is_master;
+    firstChildNodes['parent'] = realm ? realm.name : '';
+    return firstChildNodes;
+  }
+
+  getZoneTree(zonegroup: RgwZonegroup, zone: any, realm?: RgwRealm) {
+    let secondChildNodes = {};
+    secondChildNodes['id'] = zone.id;
+    secondChildNodes['name'] = zone.name + ' (zone)';
+    secondChildNodes['info'] = zone;
+    secondChildNodes['parent'] = zonegroup.name;
+    secondChildNodes['second_parent'] = realm ? realm.name : '';
+    secondChildNodes['is_default'] = zone.name === this.defaultsInfo['default_zone'] ? true : false;
+    secondChildNodes['is_master'] = zonegroup.master_zone === zone.id ? true : false;
+    return secondChildNodes;
+  }
+
+  private abstractTreeData(data: any): any[] {
```

Comment:
```suggestion
  private abstractTreeData(data: any): any[] {
```
I'm a bit concerned by the lack of typing here... Can't we be more specific with the types of the input and output data?